collector/cluster_license.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type clusterLicenseMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(clusterLicenseStats clusterLicenseResponse) float64
+	Labels func(clusterLicenseStats clusterLicenseResponse) []string
+}
+
+var (
+	defaultClusterLicenseLabels = []string{"cluster_license_type"}
+	defaultClusterLicenseValues = func(clusterLicense clusterLicenseResponse) []string {
+		return []string{clusterLicense.License.Type}
+	}
+)
+
+// License Information Struct
+type ClusterLicense struct {
+	logger log.Logger
+	client *http.Client
+	url    *url.URL
+
+	clusterLicenseMetrics []*clusterLicenseMetric
+}
+
+// NewClusterLicense defines ClusterLicense Prometheus metrics
+func NewClusterLicense(logger log.Logger, client *http.Client, url *url.URL) *ClusterLicense {
```

Comment:
This should use the collector interface. This is the current style of collectors that we are supporting. The best example of this is the snapshot collector and you can find the PR #789 showing the work to use the new style. A few notes about matching the new style:
- The `NewClusterLicense` func should return a Collector and an error
- There should not be a slice of metrics (i.e. `[]*clusterLicenseMetric).  In fact the clusterLicenseMetric type isn't necessary at all. Instead, define each metric as a var at the package level. This allows linters to warn when metrics are not referenced.
- We don't need default labels/values. When you remove the clusterLicenseMetric struct, they will no longer serve a purpose. 
-  fetchAndDecode logic should be incorporated into the `Update()` func. There's no need to separate it out.